### PR TITLE
Remove usage of merkledag.Link.Node pointer outside of merkledag

### PR DESCRIPTION
**Pull request by mildred**
This prepares for inclusion of IPLD where the Node pointer won't be there. In any case, we should probably move away from using the struct attributes and use functions to access them instead. This will make it easier to plug in IPLD in place of merkledag.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 38734c7d1cf1923e2774ae525396e65c87c9cd77
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

File: merkledag/node.go [hunk 1]
```diff
@@ -83,6 +84,26 @@ func (l *Link) GetNode(ctx context.Context, serv DAGService) (*Node, error) {
 	return serv.Get(ctx, key.Key(l.Hash))
 }
 
+// GetNodeAndCache return the MDAG Node that the link points to and store a
+// pointer to that node along with the link to speed up further retrivals. A
+// timeout is to be specified to avoid taking too much time.
+func (l *Link) GetNodeAndCache(ctx context.Context, serv DAGService, timeout time.Duration) (*Node, error) {
```
Comment (whyrusleeping) [inline merkledag/node.go]: timeout should be specified on the context using `context.WithTimeout` on the caller.

**Comment by mildred** [comment]
**Note:** We have two ways to go now:

- Make go-ipld gold a pointer to the Node object. It won't be as straightforward, and that's why we must limit usage of that pointer to the merkledag package (and have other packages uses functions instead)

- Remove the notion that a merkledag.Node can hold a pointer to its linked nodes, and refactor the API to reflect that.

    If we decide to refactor the API, the following functions will need to be modified:

    - `DAGService.AddRecursive(Node)`
    - `DAGService.Remove(Node)`
    - `Node.AddNodeLink(string, Node)`
    - `Node.AddRawLink(string, Link)`
    - `Node.UpdateNodeLink(string, Link)`: Only the codehttp PUT handler uses it, and this is being refactored in PR #1886
    - `Link.GetNode(...)` : has the ability to query the DAGService every time if the Node pointer is not there (may be slower, how much?)
    - `Node.GetLinkedNode(..., string)` : same as above

    A simple approach could be to introduce a `NodeTree` type that will hold a root node and a collection of `Node` dependencies. When constructing a DAG, we will use the `NodeTree` type instead of `Node`. `AddRecursive(Node)` will have to be renames `AddTree(NodeTree)`.

**Comment by jbenet** [comment]
let's do this on top of `dev0.4.0` branch

**Comment by jbenet** [comment]
see also discussion in https://github.com/ipfs/go-ipld/issues/14

**Pull request closed without merging**